File: core/src/main/java/aurum/lang/model/attribute/OperatorAttribute.java
package aurum.lang.model.attribute;

import aurum.lang.model.Attribute;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

public class OperatorAttribute implements Attribute {
    private final String symbol;
    private final int precedence;
    private final Associativity associativity;
    private final OperatorType type;

    public OperatorAttribute(String symbol, int precedence, Associativity associativity, OperatorType type) {
        this.symbol = symbol;
        this.precedence = precedence;
        this.associativity = associativity;
        this.type = type;
    }

    @Override
    public @NotNull String name() {
        return "Operator";
    }

    @Override
    public @NotNull Map<@NotNull String, Object> values() {
        return Map.of(
                "symbol", symbol,
                "precedence", precedence,
                "associativity", associativity,
                "type", type
        );
    }

    public enum Associativity {
        LEFT_TO_RIGHT, RIGHT_TO_LEFT
    }

    public enum OperatorType {
        BINARY, UNARY
    }
}

File: ir/src/main/kotlin/lang/aurum/attribute/OperatorAttribute.kt
package lang.aurum.attribute

import lang.aurum.ir.Associativity
import lang.aurum.ir.Opcode
import lang.aurum.ir.Operator
import lang.aurum.model.Attribute

data class OperatorAttribute (
    val operator: Operator
) : Attribute {
    val values: MutableMap<String, Any?> = mutableMapOf(
        "symbol" to operator.symbol,
        "precedence" to operator.precedence,
        "associativity" to operator.associativity,
    )

    override fun name(): String {
        return "Operator"
    }

    override fun values(): MutableMap<String, Any?> {
        return values
    }

    val symbol: String = operator.symbol
    val precedence: Int = operator.precedence
    val associativity: Associativity = operator.associativity
    val isBinary: Boolean = operator.isBinary
    val defaultOpcode: Opcode? = operator.defaultOpcode
}
